src/Compilers/CSharp/Test/Symbol/Symbols/ModuleInitializersTests.Targets.cs

```diff
@@ -0,0 +1,508 @@
+            compilation.VerifyEmitDiagnostics(
+                // (6,6): error CS8795: A module initializer must be an ordinary member method
+                //     [ModuleInitializer]
+                Diagnostic(ErrorCode.ERR_ModuleInitializerMethodMustBeOrdinary, "ModuleInitializer").WithLocation(6, 6)
+                );
+        }
+
+        [Fact]
+        public void IgnoredOnStaticConstructorWithBadAttributeTargets()
+        {
+            string source = @"
+using System;
+using System.Runtime.CompilerServices;
+
+class C
+{
+    [ModuleInitializer]
+    static C() { }
+}
+
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Method)]
+    class ModuleInitializerAttribute : System.Attribute { }
+}
+";
+            var compilation = CreateCompilation(source, parseOptions: s_parseOptions);
+            compilation.VerifyEmitDiagnostics(
+                // (7,6): error CS0592: Attribute 'ModuleInitializer' is not valid on this declaration type. It is only valid on 'method' declarations.
+                //     [ModuleInitializer]
+                Diagnostic(ErrorCode.ERR_AttributeOnBadSymbolType, "ModuleInitializer").WithArguments("ModuleInitializer", "method").WithLocation(7, 6)
+                );
+        }
+
+        [Fact]
+        public void TargetMustNotBeInstanceConstructor()
+        {
+            string source = @"
+using System.Runtime.CompilerServices;
+
+class C
+{
+    [ModuleInitializer]
+    public C() { }
+}
+
+namespace System.Runtime.CompilerServices { class ModuleInitializerAttribute : System.Attribute { } }
+";
+            var compilation = CreateCompilation(source, parseOptions: s_parseOptions);
+            compilation.VerifyEmitDiagnostics(
+                // (6,6): error CS8795: A module initializer must be an ordinary member method
+                //     [ModuleInitializer]
+                Diagnostic(ErrorCode.ERR_ModuleInitializerMethodMustBeOrdinary, "ModuleInitializer").WithLocation(6, 6)
+                );
+        }
+
+        [Fact]
+        public void IgnoredOnInstanceConstructorWithBadAttributeTargets()
+        {
+            string source = @"
+using System;
+using System.Runtime.CompilerServices;
+
+class C
+{
+    [ModuleInitializer]
+    public C() { }
+}
+
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Method)]
+    class ModuleInitializerAttribute : System.Attribute { }
+}
+";
+            var compilation = CreateCompilation(source, parseOptions: s_parseOptions);
+            compilation.VerifyEmitDiagnostics(
+                // (7,6): error CS0592: Attribute 'ModuleInitializer' is not valid on this declaration type. It is only valid on 'method' declarations.
+                //     [ModuleInitializer]
+                Diagnostic(ErrorCode.ERR_AttributeOnBadSymbolType, "ModuleInitializer").WithArguments("ModuleInitializer", "method").WithLocation(7, 6)
+                );
+        }
+    }
+}
```

Comment:
Consider testing what happens when the target of the `[ModuleInitializerAttribute]` is the definition part of a `static partial` method with no implementation part.

Consider testing what happens when the target of the `[ModuleInitializerAttribute]` is the implementation part of a `static partial` method.

Reply:
Yes, though public partial methods in language design right now IIRC. Since there wasn't anything to test though, I skipped it.